pkg/traceql/ast_metrics.go

```diff
@@ -0,0 +1,360 @@
+	return v, t
+}
+
+// exemplarAttribute captures a closure around the attribute so it doesn't have to be passed along with every span.
+// should be more efficient.
+func exemplarAttribute(a Attribute) func(Span) (float64, uint64) {
+	return func(s Span) (float64, uint64) {
+		v, _ := FloatizeAttribute(s, a)
+		t := s.StartTimeUnixNanos() / uint64(time.Millisecond)
+		return v, t
+	}
+}
+
+func (a *MetricsAggregate) initSum(q *tempopb.QueryRangeRequest) {
+	// Currently all metrics are summed by job to produce
+	// intermediate results. This will change when adding min/max/topk/etc
+	a.seriesAgg = NewSimpleCombiner(q, a.simpleAggregationOp)
+}
+
+func (a *MetricsAggregate) initFinal(q *tempopb.QueryRangeRequest) {
+	switch a.op {
+	case metricsAggregateQuantileOverTime:
+		a.seriesAgg = NewHistogramAggregator(q, a.floats)
+	default:
+		// These are simple additions by series
+		a.seriesAgg = NewSimpleCombiner(q, a.simpleAggregationOp)
+	}
+}
+
+func (a *MetricsAggregate) observe(span Span) {
+	a.agg.Observe(span)
+}
+
+func (a *MetricsAggregate) observeExemplar(span Span) {
+	v, ts := a.exemplarFn(span)
+	a.agg.ObserveExemplar(span, v, ts)
+}
+
+func (a *MetricsAggregate) observeSeries(ss []*tempopb.TimeSeries) {
+	a.seriesAgg.Combine(ss)
+}
+
+func (a *MetricsAggregate) result() SeriesSet {
+	if a.agg != nil {
+		return a.agg.Series()
+	}
+
+	// In the frontend-version the results come from
+	// the job-level aggregator
+	return a.seriesAgg.Results()
+}
+
+func (a *MetricsAggregate) validate() error {
+	switch a.op {
+	case metricsAggregateCountOverTime:
+	case metricsAggregateMinOverTime:
+	case metricsAggregateMaxOverTime:
+	case metricsAggregateRate:
+	case metricsAggregateHistogramOverTime:
+		if len(a.by) >= maxGroupBys {
+			// We reserve a spot for the bucket so quantile has 1 less group by
+			return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+		}
+	case metricsAggregateQuantileOverTime:
+		if len(a.by) >= maxGroupBys {
+			// We reserve a spot for the bucket so quantile has 1 less group by
+			return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+		}
+		for _, q := range a.floats {
+			if q < 0 || q > 1 {
+				return fmt.Errorf("quantile must be between 0 and 1: %v", q)
+			}
+		}
+	default:
+		return newUnsupportedError(fmt.Sprintf("metrics aggregate operation (%v)", a.op))
+	}
+
+	if len(a.by) > maxGroupBys {
+		return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+	}
+
+	return nil
+}
+
+var _ metricsFirstStageElement = (*MetricsAggregate)(nil)
+
+// metricsSecondStageElement represents operations that can be performed
+// after the first stage metrics pipeline, such as topK/bottomK, etc.
+type metricsSecondStageElement interface {
+	Element
+	process(input SeriesSet) SeriesSet
+}
+
+// MetricsSecondStage handles second stage metrics operations (topK/bottomK)
+type MetricsSecondStage struct {
```

Comment:
The metrics first stage elements all shared the same struct, and now I wish we had broken them out into dedicated structs.  Let's go ahead and break these out into separate Topk/BottomKElement struct and `newTopK()/newBottomK()` funcs?  Or at least rename to something like `TopKBottomKElement` so that we aren't tempted to make another mega struct for all second stage metrics elements.

Reply:
done